README.md

 --->>> Wiki


changelog.txt
[B]Version 1.2.4 - 2016-03-30[/B]
- Bugfix. Missing logo will now be replaced by fallback-logo
- Bugfix. ottkey or placeholder within M3U list will now be replaced by customer key. Before-> "http://ott.watch/XXXXXXXX/1.m3u", after-> "http://ott.watch/CUSTOMERKEY/1.m3u"
[B]Version 1.2.2 - 2016-03-30[/B]
- EPG und Archivzeiten wurden auf UTC umgestellt


addon.py
#
#      Copyright (C) 2012 Tommy Winther
#      http://tommy.winther.nu
#
#  This Program is free software; you can redistribute it and/or modify
#  it under the terms of the GNU General Public License as published by
#  the Free Software Foundation; either version 2, or (at your option)
#  any later version.
#
#  This Program is distributed in the hope that it will be useful,
#  but WITHOUT ANY WARRANTY; without even the implied warranty of
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
#  GNU General Public License for more details.
#
#  You should have received a copy of the GNU General Public License
#  along with this Program; see the file LICENSE.txt.  If not, write to
#  the Free Software Foundation, 675 Mass Ave, Cambridge, MA 02139, USA.
#  http://www.gnu.org/copyleft/gpl.html
#
import buggalo
import gui

#import pydevd
#pydevd.settrace(stdoutToServer=True, stderrToServer=True)

buggalo.SUBMIT_URL = 'http://noname.com/exception/submit.php'

try:
    w = gui.TVGuide()
    w.doModal()
    del w

except Exception:
    buggalo.onExceptionRaised()


service.py
#
#      Copyright (C) 2012 Tommy Winther
#      http://tommy.winther.nu
#
#  This Program is free software; you can redistribute it and/or modify
#  it under the terms of the GNU General Public License as published by
#  the Free Software Foundation; either version 2, or (at your option)
#  any later version.
#
#  This Program is distributed in the hope that it will be useful,
#  but WITHOUT ANY WARRANTY; without even the implied warranty of
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
#  GNU General Public License for more details.
#
#  You should have received a copy of the GNU General Public License
#  along with this Program; see the file LICENSE.txt.  If not, write to
#  the Free Software Foundation, 675 Mass Ave, Cambridge, MA 02139, USA.
#  http://www.gnu.org/copyleft/gpl.html
#
import xbmcaddon
import notification
import xbmc
import source
import time


class Service(object):
    def __init__(self):
        self.database = source.Database()
        self.database.initialize(self.onInit)

    def onInit(self, success):
        if success:
            self.database.updateChannelAndProgramListCaches(self.onCachesUpdated)
        else:
            self.database.close()

    def onCachesUpdated(self):

        if ADDON.getSetting('notifications.enabled') == 'true':
            n = notification.Notification(self.database, ADDON.getAddonInfo('path'))
            n.scheduleNotifications()

        self.database.close(None)


if __name__ == '__main__':
    monitor = xbmc.Monitor()
    
    #Service ausfuehren 
    try:
        ADDON = xbmcaddon.Addon(id = 'script.tvglobus')
        if ADDON.getSetting('cache.data.on.xbmc.startup') == 'true':
            Service()
    except source.SourceNotConfiguredException:
        pass  # ignore
    except Exception, ex:
        xbmc.log('[script.tvglobus] Uncaugt exception in service.py: %s' % str(ex) , xbmc.LOGDEBUG)
    
    while not monitor.abortRequested():
        # Sleep/wait for abort for 10 seconds
        if monitor.waitForAbort(10):
            # Abort was requested while waiting. We should exit
            break
        xbmc.log("Monitor Thread: waiting for kodis end -> %s" % time.time(), level=xbmc.LOGDEBUG)
        

#try:
#    ADDON = xbmcaddon.Addon(id = 'script.tvglobus')
#    if ADDON.getSetting('cache.data.on.xbmc.startup') == 'true':
#        Service()
#except source.SourceNotConfiguredException:
#    pass  # ignore
#except Exception, ex:
#    xbmc.log('[script.tvglobus] Uncaugt exception in service.py: %s' % str(ex) , xbmc.LOGDEBUG)


strings.py
#
#      Copyright (C) 2012 Tommy Winther
#      http://tommy.winther.nu
#
#  This Program is free software; you can redistribute it and/or modify
#  it under the terms of the GNU General Public License as published by
#  the Free Software Foundation; either version 2, or (at your option)
#  any later version.
#
#  This Program is distributed in the hope that it will be useful,
#  but WITHOUT ANY WARRANTY; without even the implied warranty of
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
#  GNU General Public License for more details.
#
#  You should have received a copy of the GNU General Public License
#  along with this Program; see the file LICENSE.txt.  If not, write to
#  the Free Software Foundation, 675 Mass Ave, Cambridge, MA 02139, USA.
#  http://www.gnu.org/copyleft/gpl.html
#
import xbmcaddon

ADDON = xbmcaddon.Addon(id = 'script.tvglobus')

NO_DESCRIPTION = 30000
CALCULATING_REMAINING_TIME = 30002
TIME_LEFT = 30003
BACKGROUND_UPDATE_IN_PROGRESS = 30004

NO_PROGRAM_AVAILABLE = 30009

NO_STREAM_AVAILABLE_TITLE = 30100
NO_STREAM_AVAILABLE_LINE1 = 30101
NO_STREAM_AVAILABLE_LINE2 = 30102

CLEAR_CACHE = 30104
CLEAR_NOTIFICATIONS = 30108
DONE = 30105
CLEAR_DBDATA = 30136

LOAD_ERROR_TITLE = 30150
LOAD_ERROR_LINE1 = 30151
LOAD_ERROR_LINE2 = 30152
CONFIGURATION_ERROR_LINE2 = 30153

SKIN_ERROR_LINE1 = 30154
SKIN_ERROR_LINE2 = 30155
SKIN_ERROR_LINE3 = 30156

NOTIFICATION_5_MINS = 30200
NOTIFICATION_NOW = 30201

WATCH_CHANNEL = 30300
REMIND_PROGRAM = 30301
DONT_REMIND_PROGRAM = 30302
CHOOSE_STRM_FILE = 30304
REMOVE_STRM_FILE = 30306

PREVIEW_STREAM = 30604
STOP_PREVIEW = 30607

WEEBTV_WEBTV_MISSING_1 = 30802
WEEBTV_WEBTV_MISSING_2 = 30803
WEEBTV_WEBTV_MISSING_3 = 30804

DATABASE_SCHEMA_ERROR_1 = 30157
DATABASE_SCHEMA_ERROR_2 = 30158
DATABASE_SCHEMA_ERROR_3 = 30159

def strings(id, replacements = None):
    string = ADDON.getLocalizedString(id)
    if replacements is not None:
        return string % replacements
    else:
        return string

dbclearer.py
#
#      Copyright (C) 2012 Tommy Winther
#      http://tommy.winther.nu
#
#  This Program is free software; you can redistribute it and/or modify
#  it under the terms of the GNU General Public License as published by
#  the Free Software Foundation; either version 2, or (at your option)
#  any later version.
#
#  This Program is distributed in the hope that it will be useful,
#  but WITHOUT ANY WARRANTY; without even the implied warranty of
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
#  GNU General Public License for more details.
#
#  You should have received a copy of the GNU General Public License
#  along with this Program; see the file LICENSE.txt.  If not, write to
#  the Free Software Foundation, 675 Mass Ave, Cambridge, MA 02139, USA.
#  http://www.gnu.org/copyleft/gpl.html
#
import datetime
import os
import xbmc
import xbmcgui
import source as src

from strings import *

class DBClear(object):
    
#    def __init__(self):
#        self.logopath = 'a'

    def onDBdataCleared(self):
        xbmcgui.Dialog().ok(strings(CLEAR_DBDATA), strings(DONE))

    def clearAllData(self):
        profilePath = xbmc.translatePath(ADDON.getAddonInfo('profile'))
        if not os.path.exists(profilePath):
            os.makedirs(profilePath)
        self.databasePath = os.path.join(profilePath, 'source.db')
        os.remove(self.databasePath)


if __name__ == '__main__':
    clearobject=DBClear()
    clearobject.clearAllData()
    clearobject.onDBdataCleared()


m3uparserutil.py
# more info on the M3U file format available here:
# http://n4k3d.com/the-m3u-file-format/

import sys
import urllib2
import StringIO
import json
import xbmc

class Track:
    def __init__(self, length, title, path, chid, chottid, logo, serverurl):
        self.length = length
        self.title = title
        self.path = path
        self.chid = chid
        self.chottid = chottid
        self.logo=logo
        self.serverurl=serverurl

class OTTLogo:
    def __init__(self, chottid, logopath):
        self.logopath = logopath
        self.chottid = chottid
        
class M3UParser:
    
    def getOTTLogos(self,logourl):
        response = urllib2.urlopen(logourl)
        #jsondata=json.loads(response.channel_now)
        
        logojson = StringIO.StringIO()
        logojson.write(response.read())
        logojson.seek(0)
        response.close()
        #js=logojson.read().split("(")[1].strip(")")
        js=logojson.read()
        js=js[js.find('(')+1:].rstrip()[:-1]

        return json.loads(js)
    
    # # # song info lines are formatted like:
    #EXTINF:419,Alice In Chains - Rotten Apple
    # length (seconds)
    # Song title
    # # # file name - relative or absolute path of file
    # ..\Minus The Bear - Planet of Ice\Minus The Bear_Planet of Ice_01_Burying Luck.mp3
    def parseM3U(self,infile,serverurl,ottkey):
        #inf = open(infile,'r')
        xbmc.log(msg='m3uparse: open file:', level=xbmc.LOGDEBUG)
        inf = infile
    
        # # # all m3u files should start with this line:
        #EXTM3U
        # this is not a valid M3U and we should stop..
        line = inf.readline()
        xbmc.log(msg='m3uparse: firstline:'+str(line), level=xbmc.LOGDEBUG)
        # initialize playlist variables before reading file
        playlist=[]
        if '#EXTM3U' not in line:
            #line.startswith('#EXTM3U'):
            xbmc.log(msg='m3uparse: return wird ausgeloesst:', level=xbmc.LOGDEBUG)
            return       
        xbmc.log(msg=str('m3uparse: get logos from : '+str(serverurl)+'/api/channel_now.jsonp'), level=xbmc.LOGDEBUG)
        logos=self.getOTTLogos(serverurl+'/api/channel_now.jsonp')
        xbmc.log(msg='m3uparse: logo:'+str(logos), level=xbmc.LOGDEBUG)
        song=Track(None,None,None,None,None,None,None)
        chid=20000
        for line in inf:
            line=line.strip()
            #xbmc.log(msg='m3uparse: line:'+str(line), level=xbmc.LOGDEBUG)
            if line.startswith('#EXTINF:'):
                # pull length and title from #EXTINF line
                length,title=line.split('#EXTINF:')[1].split(',',1)
                xbmc.log(msg='m3uparse: title:'+str(title), level=xbmc.LOGDEBUG)
                xbmc.log(msg='m3uparse: length:'+str(length), level=xbmc.LOGDEBUG)
                # title=title.decode('UTF-8')
                song=Track(length,title,None,None,None,None,None)
                xbmc.log(msg='m3uparse: Track saved with length and title', level=xbmc.LOGDEBUG)
            elif (len(line) != 0):
                xbmc.log(msg='m3uparse: add element to Playlist', level=xbmc.LOGDEBUG)
                # pull song path from all other, non-blank lines
                #song.path=line
                #xbmc.log(msg='m3uparse: ottkey: '+ottkey, level=xbmc.LOGDEBUG)
                song.path=self.replaceOTTKey(line,ottkey)
                xbmc.log(msg='m3uparse: add element-url: '+song.path, level=xbmc.LOGDEBUG)
                playlist.append(song)
                lastslash=song.path.rfind('/')
                lastpoint=song.path.rfind('.')
                song.chottid=song.path[lastslash+1:lastpoint:]
                xbmc.log(msg='m3uparse: playlist element shottid:'+str(song.chottid), level=xbmc.LOGDEBUG)
                chid+=1
                song.chid=chid
                xbmc.log(msg='m3uparse: playlist element serverurl: '+serverurl, level=xbmc.LOGDEBUG)
                song.serverurl=serverurl
                try:
                    song.logo=song.serverurl+'/images/'+logos[song.chottid]['img']
                    xbmc.log(msg='m3uparse: playlist element logo: '+song.serverurl+'/images/'+logos[song.chottid]['img'], level=xbmc.LOGDEBUG)
                except:
                    song.logo='https://upload.wikimedia.org/wikipedia/en/c/c6/TV_Guide_logo_2016.png' 
                    xbmc.log(msg='m3uparse: playlist element logo: FALLBACK_LOGO', level=xbmc.LOGDEBUG)
                # reset the song variable so it doesn't use the same EXTINF more than once
                song=Track(None,None,None,None,None,None,None)
                xbmc.log(msg='m3uparse: Track saved with length and title', level=xbmc.LOGDEBUG)
    
        inf.close()
    
        return playlist

    def replaceOTTKey(self,oldpath,newOTTKey):
        parts= oldpath.split('/')
#        print 'PARTTTTTTT 0:'+parts[0]
#        print 'PARTTTTTTT 1:'+parts[1]
#        print 'PARTTTTTTT 2:'+parts[2]
#        print 'PARTTTTTTT 3:'+parts[3]
#        print 'PARTTTTTTT 4:'+parts[4]
#        print 'PARTTTTTTT 5:'+parts[5]
        if len(parts)==6:
            return parts[0]+'//'+parts[2]+'/'+parts[3]+'/'+newOTTKey+'/'+parts[5]
        else:
            return oldpath

# for now, just pull the track info and print it onscreen
# get the M3U file path from the first command line argument
#def main():
#    m3ufile=sys.argv[1]
#    playlist = M3UParser.parseM3U(m3ufile)
#    for track in playlist:
#        print (track.title, track.length, track.path)

#if __name__ == '__main__':
#    main()

notification.py
#
#      Copyright (C) 2012 Tommy Winther
#      http://tommy.winther.nu
#
#  This Program is free software; you can redistribute it and/or modify
#  it under the terms of the GNU General Public License as published by
#  the Free Software Foundation; either version 2, or (at your option)
#  any later version.
#
#  This Program is distributed in the hope that it will be useful,
#  but WITHOUT ANY WARRANTY; without even the implied warranty of
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
#  GNU General Public License for more details.
#
#  You should have received a copy of the GNU General Public License
#  along with this Program; see the file LICENSE.txt.  If not, write to
#  the Free Software Foundation, 675 Mass Ave, Cambridge, MA 02139, USA.
#  http://www.gnu.org/copyleft/gpl.html
#
import datetime
import os
import xbmc
import xbmcgui
import source as src

from strings import *


class Notification(object):
    def __init__(self, database, addonPath):
        """
        @param database: source.Database
        """
        self.database = database
        self.addonPath = addonPath
        self.icon = os.path.join(self.addonPath, 'icon.png')

    def createAlarmClockName(self, programTitle, startTime):
        return 'tvguide-%s-%s' % (programTitle, startTime)

    def scheduleNotifications(self):
        xbmc.log("[script.tvglobus] Scheduling notifications")
        for channelTitle, programTitle, startTime in self.database.getNotifications():
            self._scheduleNotification(channelTitle, programTitle, startTime)

    def _scheduleNotification(self, channelTitle, programTitle, startTime):
        t = startTime - datetime.datetime.now()
        timeToNotification = ((t.days * 86400) + t.seconds) / 60
        if timeToNotification < 0:
            return

        name = self.createAlarmClockName(programTitle, startTime)

        description = strings(NOTIFICATION_5_MINS, channelTitle)
        xbmc.executebuiltin('AlarmClock(%s-5mins,Notification(%s,%s,10000,%s),%d,True)' %
            (name.encode('utf-8', 'replace'), programTitle.encode('utf-8', 'replace'), description.encode('utf-8', 'replace'), self.icon, timeToNotification - 5))

        description = strings(NOTIFICATION_NOW, channelTitle)
        xbmc.executebuiltin('AlarmClock(%s-now,Notification(%s,%s,10000,%s),%d,True)' %
                            (name.encode('utf-8', 'replace'), programTitle.encode('utf-8', 'replace'), description.encode('utf-8', 'replace'), self.icon, timeToNotification))

    def _unscheduleNotification(self, programTitle, startTime):
        name = self.createAlarmClockName(programTitle, startTime)
        xbmc.executebuiltin('CancelAlarm(%s-5mins,True)' % name.encode('utf-8', 'replace'))
        xbmc.executebuiltin('CancelAlarm(%s-now,True)' % name.encode('utf-8', 'replace'))

    def addNotification(self, program):
        self.database.addNotification(program)
        self._scheduleNotification(program.channel.title, program.title, program.startDate)

    def removeNotification(self, program):
        self.database.removeNotification(program)
        self._unscheduleNotification(program.title, program.startDate)


if __name__ == '__main__':
    database = src.Database()

    def onNotificationsCleared():
        xbmcgui.Dialog().ok(strings(CLEAR_NOTIFICATIONS), strings(DONE))

    def onInitialized(success):
        if success:
            database.clearAllNotifications()
            database.close(onNotificationsCleared)
        else:
            database.close()

    database.initialize(onInitialized)
